File: plugins/modulation/rotary-speaker/RotarySpeaker.java
package com.vocalmonitor.plugin.community;

import com.vocalmonitor.plugin.VocalMonitorNativePlugin;

// Rotary Speaker (native port) — Leslie cabinet sim. Crossover splits
// the signal into bass-rotor drum (lows) and horn-rotor treble (highs);
// each rotor adds Doppler delay modulation + amplitude tremolo at its
// own LFO rate. Slow/fast switching ramps like real motor inertia.
public final class RotarySpeaker implements VocalMonitorNativePlugin {
    private float lpState = 0f;
    private float hpPrev = 0f;
    private float hpOut = 0f;
    private float[] bassBuf;
    private float[] hornBuf;
    private int bassLen, hornLen;
    private int bassIdx = 0, hornIdx = 0;
    private float bassPhase = 0f;
    private float hornPhase = (float) Math.PI;
    private float bassRate = 0.8f;
    private float hornRate = 1.0f;
    private int sampleRate = 44100;
    private float speedKnob = 0f, depth = 0.7f, drive = 0.2f, mix = 1f;

    @Override
    public void init(int sr) {
        this.sampleRate = sr;
        bassLen = (int) Math.floor(sr * 0.03);
        hornLen = (int) Math.floor(sr * 0.02);
        bassBuf = new float[bassLen];
        hornBuf = new float[hornLen];
        bassIdx = hornIdx = 0;
        bassPhase = 0f;
        hornPhase = (float) Math.PI;
        bassRate = 0.8f; hornRate = 1.0f;
        lpState = hpPrev = hpOut = 0f;
    }

    @Override public String[] parameterNames() { return new String[] { "speed", "depth", "drive", "mix" }; }
    @Override public float parameterMin(String n) { return 0f; }
    @Override public float parameterMax(String n) { return 1f; }
    @Override public float parameterDefault(String n) {
        switch (n) {
            case "speed": return 0f;
            case "depth": return 0.7f;
            case "drive": return 0.2f;
            case "mix":   return 1f;
            default:      return 0f;
        }
    }
    @Override public String parameterLabel(String n) {
        switch (n) {
            case "speed": return "Speed";
            case "depth": return "Depth";
            case "drive": return "Drive";
            case "mix":   return "Mix";
            default:      return n;
        }
    }
    @Override public void setParameter(String n, float v) {
        switch (n) {
            case "speed": speedKnob = v; break;
            case "depth": depth = v; break;
            case "drive": drive = v; break;
            case "mix":   mix = v; break;
        }
    }

    @Override
    public void process(float[] input, float[] output) {
        final float targetBass = 0.7f + speedKnob * 5.8f;
        final float targetHorn = 1.0f + speedKnob * 6.5f;
        final float rampCoef = 1f - (float) Math.exp(-1.0 / Math.max(1, sampleRate * 0.6));
        final float dt = 1f / sampleRate;
        final float rcLp = (float) (1.0 / (2.0 * Math.PI * 800.0));
        final float lpA = dt / (rcLp + dt);
        final float hpRc = (float) (1.0 / (2.0 * Math.PI * 200.0));
        final float hpA = hpRc / (hpRc + dt);
        final float bassMaxDepth = sampleRate * 0.004f;
        final float hornMaxDepth = sampleRate * 0.0015f;
        final float driveK = 1f + drive * 5f;
        final float driveNorm = 1f / (float) Math.tanh(driveK);
        final float depthLocal = depth;
        final float mixLocal = mix;
        final float dry = 1f - mixLocal;
        final float twoPi = (float) (2.0 * Math.PI);
        final float invSr = 1f / sampleRate;
        final float[] bb = bassBuf, hb = hornBuf;
        final int bL = bassLen, hL = hornLen;
        int bIdx = bassIdx, hIdx = hornIdx;
        float bP = bassPhase, hP = hornPhase;
        float bR = bassRate, hR = hornRate;
        float lpS = lpState, hpP = hpPrev, hpO = hpOut;

        final int n = input.length;
        for (int i = 0; i < n; i++) {
            float x = input[i];
            float driven = (float) Math.tanh(x * driveK) * driveNorm;
            lpS = lpS + lpA * (driven - lpS);
            float lows = lpS;
            hpO = hpA * (hpO + driven - hpP);
            hpP = driven;
            float highs = driven - lows;

            bR = bR + rampCoef * (targetBass - bR);
            hR = hR + rampCoef * (targetHorn - hR);

            float bassLfo = (float) Math.sin(bP);
            float bassAmp = 1f - depthLocal * 0.3f * bassLfo;
            float bassDelay = bassMaxDepth * depthLocal * (1f + bassLfo) * 0.5f;
            bb[bIdx] = lows;
            float bRead = bIdx - bassDelay;
            while (bRead < 0) bRead += bL;
            int bI = (int) bRead; float bF = bRead - bI;
            int bJ = bI + 1; if (bJ >= bL) bJ = 0;
            float bassOut = (bb[bI] * (1f - bF) + bb[bJ] * bF) * bassAmp;
            bIdx++; if (bIdx >= bL) bIdx = 0;

            float hornLfo = (float) Math.sin(hP);
            float hornAmp = 1f - depthLocal * 0.45f * hornLfo;
            float hornDelay = hornMaxDepth * depthLocal * (1f + hornLfo) * 0.5f;
            hb[hIdx] = highs;
            float hRead = hIdx - hornDelay;
            while (hRead < 0) hRead += hL;
            int hI = (int) hRead; float hF = hRead - hI;
            int hJ = hI + 1; if (hJ >= hL) hJ = 0;
            float hornOut = (hb[hI] * (1f - hF) + hb[hJ] * hF) * hornAmp;
            hIdx++; if (hIdx >= hL) hIdx = 0;

            float wet = bassOut + hornOut;
            output[i] = x * dry + wet * mixLocal;

            bP += twoPi * bR * invSr; if (bP > twoPi) bP -= twoPi;
            hP += twoPi * hR * invSr; if (hP > twoPi) hP -= twoPi;
        }
        bassIdx = bIdx; hornIdx = hIdx;
        bassPhase = bP; hornPhase = hP;
        bassRate = bR; hornRate = hR;
        lpState = lpS; hpPrev = hpP; hpOut = hpO;
    }
}

File: plugins/modulation/rotary-speaker/rotary-speaker.js
// Rotary Speaker — Leslie cabinet sim. Two independently-rotating
// elements (the bass-rotor drum, the horn-rotor treble) each create
// pitch (Doppler) modulation as they swing toward and away from the
// virtual mic, plus an amplitude tremolo from the directional pattern.
// Crossover splits the signal — the bass drum gets the lows, the horn
// the highs — and each is modulated at its own rate. Two-speed switching
// (slow "chorale", fast "tremolo") with smooth ramp-up like the real
// motor inertia. The trademark Procol Harum / Hammond B-3 sound.

function RotarySpeakerProcessor() {
  // Crossover at ~800 Hz, LR2 (one-pole each).
  this.lpState = 0;
  this.hpPrev = 0;
  this.hpOut = 0;
  // Doppler delay buffers — bass slower, horn faster.
  var bassBufLen = Math.floor(sampleRate * 0.03);
  var hornBufLen = Math.floor(sampleRate * 0.02);
  this.bassBuf = new Array(bassBufLen);
  this.hornBuf = new Array(hornBufLen);
  for (var k = 0; k < bassBufLen; k++) this.bassBuf[k] = 0;
  for (var k = 0; k < hornBufLen; k++) this.hornBuf[k] = 0;
  this.bassLen = bassBufLen; this.hornLen = hornBufLen;
  this.bassIdx = 0; this.hornIdx = 0;
  this.bassPhase = 0; this.hornPhase = Math.PI; // 180° offset for stereo realism
  // Smoothed actual speeds (motor inertia).
  this.bassRate = 0.8;
  this.hornRate = 1.0;
}

RotarySpeakerProcessor.parameterDescriptors = [
  { name: 'speed',  label: 'Speed',  defaultValue: 0,   minValue: 0,  maxValue: 1 },
  { name: 'depth',  label: 'Depth',  defaultValue: 0.7, minValue: 0,  maxValue: 1 },
  { name: 'drive',  label: 'Drive',  defaultValue: 0.2, minValue: 0,  maxValue: 1 },
  { name: 'mix',    label: 'Mix',    defaultValue: 1,   minValue: 0,  maxValue: 1 }
];

RotarySpeakerProcessor.prototype.process = function (inputs, outputs, parameters) {
  var input = inputs[0][0], output = outputs[0][0];
  if (!input || !output) return true;
  var speedKnob = parameters.speed[0]; // 0 = slow, 1 = fast
  var depth = parameters.depth[0];
  var drive = parameters.drive[0];
  var mix = parameters.mix[0];

  // Target speeds. Slow ("chorale") ≈ 0.7 Hz for bass / 1.0 Hz horn.
  // Fast ("tremolo") ≈ 6.5 Hz bass / 7.5 Hz horn. Real Leslie ramps
  // between them in ~1 second — model that with a per-sample IIR.
  var targetBass = 0.7 + speedKnob * 5.8;
  var targetHorn = 1.0 + speedKnob * 6.5;
  var rampCoef = 1 - Math.exp(-1 / Math.max(1, sampleRate * 0.6));

  // One-pole crossover coefficients at ~800 Hz.
  var dt = 1 / sampleRate;
  var rcLp = 1 / (2 * Math.PI * 800);
  var lpA = dt / (rcLp + dt);
  var hpRc = 1 / (2 * Math.PI * 200);
  var hpA = hpRc / (hpRc + dt);

  var bassMaxDepth = sampleRate * 0.004; // 4 ms peak
  var hornMaxDepth = sampleRate * 0.0015; // 1.5 ms peak

  for (var i = 0; i < input.length; i++) {
    var x = input[i];

    // --- Soft preamp drive (Leslie's tube preamp). ---
    var k = 1 + drive * 5;
    var driven = Math.tanh(x * k) / Math.tanh(k);

    // --- Crossover. ---
    this.lpState = this.lpState + lpA * (driven - this.lpState);
    var lows = this.lpState;
    this.hpOut = hpA * (this.hpOut + driven - this.hpPrev);
    this.hpPrev = driven;
    var highs = driven - lows;

    // Smooth motor speed.
    this.bassRate = this.bassRate + rampCoef * (targetBass - this.bassRate);
    this.hornRate = this.hornRate + rampCoef * (targetHorn - this.hornRate);

    // --- Bass rotor: amplitude tremolo + Doppler delay modulation. ---
    var bassLfo = Math.sin(this.bassPhase);
    var bassAmp = 1 - depth * 0.3 * bassLfo;
    var bassDelay = bassMaxDepth * depth * (1 + bassLfo) * 0.5;
    this.bassBuf[this.bassIdx] = lows;
    var bRead = this.bassIdx - bassDelay;
    while (bRead < 0) bRead += this.bassLen;
    var bI = Math.floor(bRead), bF = bRead - bI;
    var bJ = bI + 1; if (bJ >= this.bassLen) bJ = 0;
    var bassOut = (this.bassBuf[bI] * (1 - bF) + this.bassBuf[bJ] * bF) * bassAmp;
    this.bassIdx++; if (this.bassIdx >= this.bassLen) this.bassIdx = 0;

    // --- Horn rotor. ---
    var hornLfo = Math.sin(this.hornPhase);
    var hornAmp = 1 - depth * 0.45 * hornLfo;
    var hornDelay = hornMaxDepth * depth * (1 + hornLfo) * 0.5;
    this.hornBuf[this.hornIdx] = highs;
    var hRead = this.hornIdx - hornDelay;
    while (hRead < 0) hRead += this.hornLen;
    var hI = Math.floor(hRead), hF = hRead - hI;
    var hJ = hI + 1; if (hJ >= this.hornLen) hJ = 0;
    var hornOut = (this.hornBuf[hI] * (1 - hF) + this.hornBuf[hJ] * hF) * hornAmp;
    this.hornIdx++; if (this.hornIdx >= this.hornLen) this.hornIdx = 0;

    var wet = bassOut + hornOut;
    output[i] = x * (1 - mix) + wet * mix;

    this.bassPhase += 2 * Math.PI * this.bassRate / sampleRate;
    if (this.bassPhase > 2 * Math.PI) this.bassPhase -= 2 * Math.PI;
    this.hornPhase += 2 * Math.PI * this.hornRate / sampleRate;
    if (this.hornPhase > 2 * Math.PI) this.hornPhase -= 2 * Math.PI;
  }
  return true;
};

registerProcessor('rotary-speaker', RotarySpeakerProcessor);
